### DISCUSSION TITLE
On the project's stability, future and funding

### DISCUSSION THREAD

**Discussion by mkurz** [apache/pekko#144] [General]
Hi everyone!

Like @gmethvin, who wrote some messages here already, I am one of the maintainers of the Play Framework.

Since Play is looking for alternatives to replace akka (including akka-http), it's obvious that pekko would be a feasible alternative for us, given it's a fork of the current akka 2.6.x branch.

We highly appreciate the ambitions you have with pekko and we sincerely hope the project succeeds in the long term.
However, we do have some concerns about pekko in the long run. That's why we haven't come up with a decision yet on how to move forward in Play.

I want express those concerns here (excuse my ignorance if some of those topics were discussed before already):

- How many seriously active and engaged devs do currently invest time to push the project forward? Specially, how much time, realistically, do individual core contributors think they can invest in the upcoming years? Please correct me if I am wrong, it seems pekko right now still is in it's early stage, being set up, but no real work has been done yet (like features and/or bugfixes).
- I assume all of you currently have full-time jobs (be it employed or as a freelancer/consulting) and some may even have families with kids. The work you do on pekko currently, is that happening outside of your regular occupation?
- Do any of you get time reserved from your employer to work on pekko? Like is your employer backing you and basically sponsoring you to work on pekko?

I do ask those questions because my fear is that in the end there will be like a handful people maintaining pekko to keep it alive in their free time, but as soon as one or two of those most engaged people struggle to invest time next to their jobs (because of whatever reasons... private life kicks in, accident, sickness, a newborn, divorce, changing jobs, single person with lots of free time not single anymore,...), the project may start to slowly fall apart. From my personal experience I know that if other stuff is coming up it sometimes gets tough to keep engaged with a project you actually only work on in your spare time. Lot's of Open Source projects actually just depend on one or two main actors that hold the project together...

That leads me to following questions:

- Have you considered getting serious funding for the pekko project, either to onboard one or two main devs, like at least part time?
- Or if one of you is freelancing/consulting anyway, have you considered working a certain amount of fixed hours for pekko as your "client" during normal working hours, not just as a side project?

Even though it took way more than half a year (almost 9 months actually) in the end we were able to get decent funding for Play, so we are now able to pay one person to work on the project. So I was wondering if you already considered opening an Open Collective and/or GitHub Sponsors account for pekko.
I do think akka is _much_ more widely used then Play ever was and believe that companies using akka right now are much more willing to fund the development of pekko than they are for Play. I also believe that around September, much more companies will feel a pressure to migrate away from akka, when 2.6 is EOL and would then be more willing to invest a bit of money into funding pekko.
Of course that only works if someone of you decides to at least work part time on the project. Or, like said, if someone of you is freelancing already anyway, you could at least charge the time to work on the pekko project, do that in normal working hours and not in the evenings or on the weekends.
If you ever choose the road of sponsorship I can offer my help to set things up.

Don't get me wrong, I am not saying funding is absolutely necessary or is the best thing to do, nor I want to talk anyone into funding, I just wanted to bring up this topic because it seems it never was discussed before.
Like I said, I am just a bit worried about the long term stability of the project.

Thanks for listening!

<details>
  <summary>PS: Side note that may be relevant for someone considering (click to open).</summary>
There are two ways to payout money from Open Collective, the standard way where you just get the money wired onto your bank account and you have to do the tax stuff yourself, nothing special, you just act like a freelancer / self-employed and do whatever you have to do yourself to pay taxes in your country.
Or, since this isn't advertised on the Open Collective website, Open Collective could also set up people on their PEO provider, as an employee with full coverage or as a contractor. In that case you would have a normal remote job, with normal regular salaries and Open Collective via its PEO provider provides you with necessary insurance and other employment records. Like I said this second option is not advertised anywhere, but if a collective has enough funding, you can just reach out to Open Collective folks to set things up. They did offer that to me, and in general, folks at Open Collective are very easy going.
Also, if you live in the European Union and need help on how to tax "income" (as freelancer/self-employed) from the Open Source Collective, which is based in the US, I can help out here too if you reach out to me.
</details>

**Comment by pjfanning** [comment]
The whole sponsorship/funding side is a bit of a grey area. ASF need [sponsorship](https://www.apache.org/foundation/thanks) to provide their services (infrastructure, security reporting, legal, events, etc.).

* In the real world, companies do hire people to work on open source and some people in the Pekko team have said that their companies are allowing them some hours to work on Pekko related work. I'll let them comment here.
* I have had contacts from 2 large companies with an interest in seeing Pekko released but did not discuss funding with them. At least with 1 of them, they are an ASF sponsor so they might feel like this is enough. There may be lots more. The FOSDEM conference at the weekend may see some interest in the Pekko stint on the Apache desk.
* It is open for people to set up their own business to support Pekko full or part time.
   * The ASF sponsoring links will have to be prominent on our web site and I'm not sure if we can also advertise sponsoring links for individual committers.
   * We might be able to add a page that lists companies active in the space.
   * Individuals could set up sponsoring links on their own Github profiles - companies/individuals looking to sponsor developer contributions should be able to find the developer accounts readily

There is a fair amount of work to do get Pekko v1.0.0 released (Akka rebadged as Pekko but with few changes from last Apache licensed releases) - and there is a little bit of pent up demand to update dependencies and fix a few bugs, so a v1.1.0 release that has some changes, that may require a little more care when uptaking, should arrive fairly quickly afterwards. Scala 3 support for pekko-http has been deferred to this v1.1.0 release, for instance, but this work is already largely done - just waiting for the conservative v1.0.0 release before it is merged into the main branch.

I haven't come across any pent up demand to do any major changes.

**Comment by luksow** [comment]
Hi,

Not sure if that's the right place but my company (iteratorshq.com) is very much interested in keeping Akka (Pekko) alive and in good condition. We heavily rely on akka-http right now and we're happy to devote some of our engineering resources (e.g. 8 hrs/week) for a more coordinated work.

**Reply by pjfanning** [reply]
Hi - thanks for the interest. https://github.com/apache/incubator-pekko-http/issues has some issues raised - feel free to take some on - might be best to comment on an issue to see if someone has started it without updating the issue - or someone might be able to highlight why the issue needs to wait (until something else is done)

**Comment by kw217** [comment]
My team are seriously considering moving to Pekko, and if we go ahead with that I expect we would have some engineering resources available to help support the project. We're unlikely to drive new features but we would absolutely have an interest in security fixes in particular.

**Comment by jona7o** [comment]
We ( @ innfactory ) use akka and play framework a lot. We hope that pekko gets the stability which is needed so that play2 will adopt it in future. Right now we just have hand of scala engineers, but if needed we are open to contribute to pekko and play2 community, because these are both great tools for modern microservices and iot plattforms written in scala. :-)

**Comment by mdedetrich** [comment]
On the funding side of things one thing that I want to explore is using https://opencollective.com/ as an official way to raise money. Initially the primary goals for this would be to procure dedicated machines both for CI and also for benchmarking but this can also be expanded in other areas if the funding gets high enough.

**Reply by pjfanning** [reply]
Can't we make a case to ASF if we need hardware or VMs. Infrastructure team are helpful in this regard? If we spec out a requirement, we may be able to advertise it if the Infra team can't help. A corporate entity or one of the big cloud providers might be able to help with VM time.

If there is money to be donated, there is an argument to focus it at developers who are giving up their free time.

**Reply by mdedetrich** [reply]
> Can't we make a case to ASF if we need hardware or VMs. Infrastructure team are helpful in this regard? If we spec out a requirement, we may be able to advertise it if the Infra team can't help.

I already did and I attended their roundup, they don't have dedicated machines and there is no desire to do so. If we want dedicated VM's we need to self host and we are all on our own.

> A corporate entity or one of the big cloud providers might be able to help with VM time.

I would prefer not to be reliant on a company for something as critical as VM's/CI. If a cloud provider that has dedicated machines can sponsor us, then great, however the whole point of https://opencollective.com/ is to solve problems like this. Companies that rely on Pekko can give us monthly donations and we can use that money to procure hardware in a completely transparent manner.

Also note there are other Apache projects on OpenCollective, its not just us.

> If there is money to be donated, there is an argument to focus it at developers who are giving up their free time.

So a dedicated machine on https://www.hetzner.com/ is around 50-70 euroes a month, we would need at most two (one for benchmarking and one for CI/multicluster tests). Thats not a lot of money for companies and there is a strong argument that using the money in this way would save a lot more developer time versus just giving it to developers (personally I would not take 50-140 euroes a month over saving many many hours on CI runs, something that is shared for all of ASF contributors).

Now if we end up getting significant amounts of money then yes this is definitely something we can consider, and the point of OpenCollective is it also solves this problem (i.e. paying OpenSource developers from donated company/user funds), thats the primary point of it.

**Reply by mkurz** [reply]
Again, if you need any help related to OpenCollective feel free to reach out to me.

**Reply by mdedetrich** [reply]
Thanks, will have capacity for this once we start getting the release going

**Comment by pjfanning** [comment]
https://apply.sovereigntechfund.de/ has funding for FOSS including under an infrastructure heading.

**Comment by pjfanning** [comment]
I created https://opencollective.com/friends-of-apache-pekko - I can add more admins if anyone is interested. I'm not really sure what content to put up but you never know - it may prove useful.

**Reply by mdedetrich** [reply]
@mkurz Pinging you explicitly because you were asking for this. Also there is a related discussion for this on https://lists.apache.org/thread/kbv10ngvjdhffc80jkf82o98ldo08ngw

**Reply by mkurz** [reply]
:+1: Like said, if you need any help just let me know.

**Comment by mdedetrich** [comment]
> I can add more admins if anyone is interested.

I would like to be an admin, I was planning to do this myself but never got around to it.

>  I'm not really sure what content to put up but you never know - it may prove useful.

I think the main thing is to add different "goals" for how the money can be used. First goals are the dedicated hardware servers for nightly runs (i.e. CI for windows/linux+linux ARM), then dedicated machines for benchmarking and then we can also have stretch goals to help pay frequent contributors (precise details is yet to be determined)

**Comment by He-Pin** [comment]
I currently can only do my contribution at my free/weekend :(, busy work days, hope we can get through all these hard.

**Comment by jxnu-liguobin** [comment]
In our company, there is no fixed time for open source support, but if necessary (bugs, enhance), it will be resolved during working hours. Unfortunately, our team has just started using pekko/akka, and only I can do it.

**Comment by laglangyue** [comment]
If there are more and more users, we will get through the difficulties. Using pekko and  contributing, for me, I may contirbute to pekko on weekend and review some PR on other day,
